velox/functions/prestosql/aggregates/MultiMapAggAggregate.cpp

```diff
@@ -0,0 +1,421 @@
+
+    decodedValues_.decode(*valueArrays_->elements());
+
+    auto accumulator = value<AccumulatorType>(group);
+
+    VELOX_CHECK_NOT_NULL(mapVector);
+    rows.applyToSelected([&](vector_size_t row) {
+      if (!decodedMaps_.isNullAt(row)) {
+        clearNull(group);
+        auto decodedRow = decodedMaps_.index(row);
+        auto offset = mapVector->offsetAt(decodedRow);
+        auto size = mapVector->sizeAt(decodedRow);
+        checkNullKeys(decodedKeys_, offset, size);
+        for (auto i = offset; i < offset + size; ++i) {
+          auto numValues = valueArrays_->sizeAt(decodedValueArrays_.index(i));
+          auto valueOffset =
+              valueArrays_->offsetAt(decodedValueArrays_.index(i));
+          accumulator->insertMultiple(
+              decodedKeys_,
+              i,
+              decodedValues_,
+              valueOffset,
+              numValues,
+              *allocator_);
+        }
+      }
+    });
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto mapVector = (*result)->as<MapVector>();
+    VELOX_CHECK(mapVector);
+    mapVector->resize(numGroups);
+    auto mapKeys = mapVector->mapKeys();
+    auto mapValueArrays = mapVector->mapValues()->as<ArrayVector>();
+    auto numKeys = countKeys(groups, numGroups);
+    mapKeys->resize(numKeys);
+    mapValueArrays->resize(numKeys);
+
+    auto mapValues = mapValueArrays->elements();
+    mapValues->resize(countValues(groups, numGroups));
+
+    auto* rawNulls = getRawNulls(mapVector);
+    vector_size_t keyOffset = 0;
+    vector_size_t valueOffset = 0;
+
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+
+      if (isNull(group)) {
+        mapVector->setNull(i, true);
+        mapVector->setOffsetAndSize(i, 0, 0);
+        continue;
+      }
+
+      clearNull(rawNulls, i);
+
+      auto accumulator = value<AccumulatorType>(group);
+      auto mapSize = accumulator->size();
+      if (mapSize) {
+        mapVector->setOffsetAndSize(i, keyOffset, mapSize);
+        accumulator->extract(mapKeys, *mapValueArrays, keyOffset, valueOffset);
+      } else {
+        mapVector->setOffsetAndSize(i, 0, 0);
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+  void destroy(folly::Range<char**> groups) override {
+    for (auto* group : groups) {
+      auto accumulator = value<AccumulatorType>(group);
+      accumulator->free(*allocator_);
+      std::destroy_at(accumulator);
+    }
+  }
+
+ private:
+  static void checkNullKeys(
+      const DecodedVector& keys,
+      vector_size_t offset,
+      vector_size_t size) {
+    static const char* kNullKey = "map key cannot be null";
+    if (keys.mayHaveNulls()) {
+      for (auto i = offset; i < offset + size; ++i) {
+        VELOX_USER_CHECK(!keys.isNullAt(i), kNullKey);
```

Comment:
Do we need add more logs here? Also checkNullKeys could be a common utility and I saw other function also have similar implementations. Consider for followup if needs. Thanks!